nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -615,12 +615,30 @@ func (w *deploymentWatcher) handleAllocUpdate(allocs []*structs.AllocListStub) (
 			continue
 		}
 
-		// Determine if the update block for this group is progress based
-		progressBased := dstate.ProgressDeadline != 0
+		hasProgressDeadline := dstate.ProgressDeadline != 0
+		failDeployment := false
```

Comment:
Because everything between line 619 and line 637 is there to set `failDeployment` with no side-effects (other than logging), it might make sense to extract that to it own function. That may make it easier to write tests of just that logic, too.